samples/bluetooth/hci_uart/src/main.c

```diff
@@ -49,78 +55,136 @@ static K_FIFO_DEFINE(tx_queue);
  */
 #define H4_DISCARD_LEN 33
 
-static int h4_read(struct device *uart, uint8_t *buf,
-		   size_t len, size_t min)
-{
-	int total = 0;
-
-	while (len) {
-		int rx;
+/* Union used for variable into which packet header is received. */
+union bt_hci_hdr {
+	struct bt_hci_cmd_hdr cmd;
+	struct bt_hci_acl_hdr acl;
+	uint8_t raw[0];
+};
 
-		rx = uart_fifo_read(uart, buf, len);
-		if (rx == 0) {
-			LOG_DBG("Got zero bytes from UART");
-			if (total < min) {
-				continue;
-			}
-			break;
-		}
+static int h4_read(struct device *uart, uint8_t *buf, size_t len)
+{
+	int rx = uart_fifo_read(uart, buf, len);
 
-		LOG_DBG("read %d remaining %d", rx, len - rx);
-		len -= rx;
-		total += rx;
-		buf += rx;
-	}
+	LOG_DBG("read %d req %d", rx, len);
 
-	return total;
+	return rx;
 }
 
-static size_t h4_discard(struct device *uart, size_t len)
+static bool valid_type(uint8_t type)
 {
-	uint8_t buf[H4_DISCARD_LEN];
-
-	return uart_fifo_read(uart, buf, MIN(len, sizeof(buf)));
+	return (type == H4_CMD) | (type == H4_ACL);
 }
 
-static void h4_cmd_recv(struct net_buf *buf, int *remaining)
+/* Function assumes that type is validated and only CMD or ACL will be used. */
+static uint32_t get_len(union bt_hci_hdr *hdr, uint8_t type)
 {
-	struct bt_hci_cmd_hdr hdr;
-
-	/* We can ignore the return value since we pass len == min */
-	h4_read(hci_uart_dev, (void *)&hdr, sizeof(hdr), sizeof(hdr));
-
-	*remaining = hdr.param_len;
-
-	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	return (type == BT_BUF_CMD) ?
+		hdr->cmd.param_len : sys_le16_to_cpu(hdr->acl.len);
+}
 
-	LOG_DBG("len %u", hdr.param_len);
+/* Function assumes that type is validated and only CMD or ACL will be used. */
+static int hdr_len(uint8_t type)
+{
+	return (type == H4_CMD) ?
+		sizeof(struct bt_hci_cmd_hdr) : sizeof(struct bt_hci_acl_hdr);
 }
 
-static void h4_acl_recv(struct net_buf *buf, int *remaining)
+static void rx_isr(void)
 {
-	struct bt_hci_acl_hdr hdr;
+	static struct net_buf *buf;
+	static int remaining;
+	static uint8_t state;
+	static uint8_t type;
+	static union bt_hci_hdr hdr;
+	int read;
+
+	do {
+		switch (state) {
+		case ST_IDLE:
+			/* Get packet type */
+			read = h4_read(hci_uart_dev, &type, sizeof(type));
+			if (read) {
+				if (valid_type(type)) {
+					/* Get expected header size and switch
+					 * to receiving header.
+					 */
+					remaining = hdr_len(type);
+					state = ST_HDR;
+				} else {
+					LOG_WRN("Unknown header %d", state);
+				}
+			}
```

Comment:
So what happens here? shouldn't we really `ASSERT(read)`? if you get an interrupt and no data, isn't that an error to be handled? Since you are only reading 1 byte here.

Reply:
Oh, I guess you are treating this as a stream. But then wouldn't it make sense to peek at `uart_fifo_read` somehow to get the total length of bytes that are available to read to avoid the `if (read)` statements?

Reply:
if rx interrupt is triggered but no data that means driver error or another user of the driver which already read that data. That is unexpected and unaccepted. Added assertion.

Reply:
yeah, we read until fifo is empty. there is no API to peek number of bytes in the fifo so i had to kept `if(read)` and added comment.